packages/cf.js/src/provider.ts

```diff
@@ -40,43 +47,150 @@ export class Provider {
   ): Promise<Node.MethodResponse> {
     const requestId = cuid();
     return new Promise<Node.MethodResponse>((resolve, reject) => {
+      const request: Node.MethodRequest = {
+        requestId,
+        params,
+        type: methodName
+      };
       this.requestListeners[requestId] = response => {
         if (response.type === Node.ErrorType.ERROR) {
-          return reject(response.data);
+          return reject({
+            type: EventType.ERROR,
+            data: response.data
+          });
         }
         if (response.type !== methodName) {
           return reject({
-            errorName: "unexpected_message_type",
-            message: `Unexpected response type. Expected ${methodName}, got ${
-              response.type
-            }`
+            type: EventType.ERROR,
+            data: {
+              errorName: "unexpected_message_type",
+              message: `Unexpected response type. Expected ${methodName}, got ${
+                response.type
+              }`
+            }
           });
         }
         resolve(response as Node.MethodResponse);
       };
       setTimeout(() => {
         if (this.requestListeners[requestId] !== undefined) {
-          reject(new Error(`Request timed out: ${requestId}`));
+          reject({
+            type: EventType.ERROR,
+            data: {
+              errorName: "request_timeout",
+              message: `Request timed out: ${JSON.stringify(request)}`
+            }
+          });
           delete this.requestListeners[requestId];
         }
       }, NODE_REQUEST_TIMEOUT);
-      this.nodeProvider.sendMessage({
-        requestId,
-        params,
-        type: methodName
-      });
+      this.nodeProvider.sendMessage(request);
     });
   }
 
   private onNodeMessage(message: Node.Message) {
-    const requestId = (message as Node.MethodResponse).requestId;
-    if (requestId) {
-      if (this.requestListeners[requestId]) {
-        this.requestListeners[requestId](message);
-        delete this.requestListeners[requestId];
-      }
+    const type = message.type;
+    if (Object.values(Node.ErrorType).indexOf(type) !== -1) {
```

Comment:
I wish there was a utility method to wrap this purely for syntactic sugar.

Though this doesn't seem like a common enough operation (yet?) to warrant that method probably.

Reply:
I did that, but it turned out to be more confusing than it's worth.